### PR TITLE
GoalDetector: Needs testing 

### PR DESCRIPTION
**Pull request by codeNinjaDev**
**Very janky code**. I struggled with differentiating three different cases for powershot detection: 

1. Camera is between Powershot and HighGoal
2. The angle to the HighGoal includes the angle to the Powershot
3. Angle to Powershot is separate from the angle to the HighGoal

In addition to the AdvancedHighGoalPipeline (Powershot calculations), I created Pipelines to track pixel values & angles of Highgoal.

image
*Case 2*
image
*Case 3*

### CODE REVIEW

**Comment by JIceberg** [comment]
How much testing? Have edge cases been looked at yet?

**Comment by codeNinjaDev** [comment]
> How much testing? Have edge cases been looked at yet?

I've tried to anticipate the above edge cases, but I haven't been able to do any testing.

**Comment by JIceberg** [comment]
This PR actually doesn't work with the refactor. Can you pull from upstream and use the new layout?

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAdvancedHighGoalPipeline.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    }
+
+    public AdvancedHighGoalPipeline(double fov, double cameraHeight, double centerOfLogoHeight, double cameraPitchOffset, double cameraYawOffset) {
+        super(fov);
+        this.cameraHeight = cameraHeight;
+        this.centerOfLogoHeight = centerOfLogoHeight;
+        this.cameraPitchOffset = cameraPitchOffset;
+        this.cameraYawOffset = cameraYawOffset;
+    }
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAdvancedHighGoalPipeline.java:30]: Constructors need to have the same name as the classfile.

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAngleHighGoalPipeline.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    public AngleHighGoalPipeline(double fov, double cameraPitchOffset, double cameraYawOffset) {
+        super();
+        this.fov = fov;
+        this.cameraPitchOffset = cameraPitchOffset;
+        this.cameraYawOffset = cameraYawOffset;
+    }
+
+    public AngleHighGoalPipeline(double fov) {
+        this(fov, 0, 0);
+    }
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAngleHighGoalPipeline.java:68]: Constructors need to have the same name as the classfile

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAngleHighGoalPipeline.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import org.opencv.core.Rect;
+import org.opencv.imgproc.Imgproc;
+import org.openftc.easyopencv.OpenCvPipeline;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+
+
+public class UGAngleHighGoalPipeline extends BasicHighGoalPipeline {
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAngleHighGoalPipeline.java:15]: Replace `BasicHighGoalPipeline` with `UGBasicHighGoalPipeline`

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAdvancedHighGoalPipeline.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
+package com.arcrobotics.ftclib.vision;
+
+import org.opencv.core.Mat;
+
+public class UGAdvancedHighGoalPipeline extends AngleHighGoalPipeline {
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGAdvancedHighGoalPipeline.java:5]: Replace `AngleHighGoalPipeline` with `UGAngleHighGoalPipeline`

File: core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGBasicHighGoalPipeline.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        biggestBlueContour = new MatOfPoint();
+        biggestRedContour = new MatOfPoint();
+
+        blueRect = new Rect();
+        redRect = new Rect();
+
+        minThreshold = 155;
+        maxThreshold = 255;
+    }
```
Comment (JIceberg) [inline core/vision/src/main/java/com/arcrobotics/ftclib/vision/UGBasicHighGoalPipeline.java:59]: Constructor should have the same name as the classfile.

**Review by JIceberg — CHANGES_REQUESTED** [review]
Follow these changes and it shouldn't fail the build check.

**Comment by codeNinjaDev** [comment]
Opening new pull request

**Pull request closed without merging**